PHAL/UART/phal_uart.c

```diff
@@ -0,0 +1,463 @@
+    assert(cb);
+    ports[port].rx_callback = cb;
+}
+
+void phal_uart_attach_transmit_cb(uart_port_id port, uartTxCallback_t cb) {
+    assert(cb);
+    ports[port].tx_callback = cb;
+}
+//
+//send byte array of len in two modes
+//(1) WAIT mode will wait for previous asynchronous send if any
+//(2) FORCE mode will pause previous asynchronous send if any
+void phal_uart_send(uart_port_id port, uint8* data, uint32 len, uart_mode mode) {
+//    while(g_TX.busy); // will wait for previous async send to finish
+    if (mode == PHAL_UART_WAIT)
+        while(ports[port].g_TX.busy);
+
+    if (ports[port].g_TX.busy)
+        ports[port].base_reg->CLEARINT = (uint32)SCI_TX_INT; //will pause any async send
+
+    while((ports[port].base_reg->FLR & (uint32)SCI_TX_INT) == 0U);
+    uint32 l;
+    for(l = 0; l < len; l++) {
+        ports[port].base_reg->TD = (uint32)data[l];
+        while((ports[port].base_reg->FLR & (uint32)SCI_TX_INT) == 0U);
+    }
+
+    if (ports[port].g_TX.busy)
+            ports[port].base_reg->SETINT = (uint32)SCI_TX_INT;
+}
+//
+////receive byte array of len in two modes
+////(1) WAIT mode will wait for previous asynchronous receive if any
+////(2) FORCE mode will pause previous asynchronous receive if any
+void phal_uart_receive(uart_port_id port, uint8* data, uint32 len, uart_mode mode) {
+    if (mode == PHAL_UART_WAIT)
+        while(ports[port].g_RX.busy);
+
+    if (ports[port].g_RX.busy)
+        ports[port].base_reg->CLEARINT = (uint32)SCI_RX_INT; //will pause any async send
+
+    uint32 l;
+    for(l = 0; l < len; l++) {
+        while((ports[port].base_reg->FLR & (uint32)SCI_RX_INT) == 0U);
+        data[l] = (uint8)(ports[port].base_reg->RD & 0xFF);
+    }
+
+    if (ports[port].g_RX.busy)
+        ports[port].base_reg->SETINT = (uint32)SCI_RX_INT; //will pause any async send
+}
+
+//
+////send byte array of len asynchronously, the flag will be set once the send is done, make sure to clear it before sending
+bool phal_uart_async_send(uart_port_id port, uint8* data, uint32 len, volatile uint32* flag) {
+    if (ports[port].intr_setup.tx_level == PHAL_UART_INTR_DISABLE)
+        return 1; //return 1 if interrupt not setup properly
+
+    if (ports[port].g_TX.busy)
+        return 1; //if not finished previous async send
+
+    if (len == 0)
+        return 0; //return immediately if length is zero
+
+    ports[port].g_TX.data = data + 1;
+    ports[port].g_TX.len = len - 1;
+    ports[port].g_TX.done = flag;
+    *ports[port].g_TX.done = 0; //clear the flag
+    ports[port].g_TX.busy = 1;
+    ports[port].base_reg->TD = data[0];
+    ports[port].base_reg->SETINT = (uint32)SCI_TX_INT;// enble TX
+    return 0;
+}
+
+//receive byte array of len asynchronously, the flag will be set once the receive is done, make sure to clear it before receiving
+bool phal_uart_async_receive(uart_port_id port, volatile uint8* data, uint32 len, volatile uint32* flag) {
+    if (ports[port].intr_setup.rx_level == PHAL_UART_INTR_DISABLE)
```

Comment:
I still don't see the point of async_send/receive. While it's true in hardware these two operations do happen asynchronously, the software still has to periodically poll for your flag and data.

Asynchronous operations in embedded systems are only effectively asynchronous if they can work independently from the processor (like PWM that toggles a pin without CPU intervention) or there is an interrupt mechanism to signal the processor that an event has occurred. 

Your API already gives the user asynchronous operation by allowing callbacks for tx/rx interrupts and if they didn't want to set that up and just poll instead, you already have normal receive and transmit functions. I don't see these functions adding any additional value.

Reply:
The async send and receive helps to keep ISRs small. It is important to keep the ISR short. This is because once the uC is in an interrupt state it will disable other interrupts. We don't want the lower priority interrupts to block the higher priority interrupts.

IRQ disables all IRQs, and FIQ disables both IRQs and FIQs.

Consider these 2 different implementations

```
void rxCallback(char data) {
  critical_task(data); // small and deterministic number of instructions
  non_critical_task(data); // large or nondeterministic number of instructions 
}

int main () {
  ... //setup uart and interrupts to receive callback every time data is received
  while(1) {
     some_other_task();
  }
  return 0;
}
```

My thinking is that this is not a good way to code ISRs, so we could alternatively implement it in this safer way.

```
void rxCallback(char data) {
  critical_task(msg); // small and deterministic number of instructions 
  async_recieve(&msg, msg_len, &flag) //start listening for next set of data (alternatively call async_read after non-critical_task)
}

int main () {
  ... //setup uart and interrupt but will not get receive interrupt until async_receive is called
 async_recieve(&msg, msg_len, &flag); // will call the rxCallback function and set the flag once it has received "msg_len" bytes of data
  while(1) {
     if (flag) { //won't block other tasks
       non_critical_task(msg); // large or nondeterministic number of instructions 
       // call async_read here instead of after critical_task (depending on the use case)
     } 
     some_other_task();
  }
  return 0
}
```

This is also easy to work when the task expects data of length more than one. This is just a simple example, the same concept can be extended to RTOS tasks. 

If there is no critical task we can avoid rxCallback altogether. The same can be done without using async functions, but I thought this will help ensure good practice. Let me know your thoughts, if you think it's unnecessary we can get rid of it.

Reply:
Ah I see, this does seem super useful for pure bare metal systems. You are referring to the practice of deferred interrupt handling. An RTOS would already provide multiple mechanisms for dealing with this though. See thsi article from the freertos website https://www.freertos.org/deferred_interrupt_processing.html

I can see how this could be useful if we decide to not use an rtos or we don't want to bother  setting up so many queues/notifications (freertos' synchronization mechanisms). I'd say there's no harm in keeping it here.

One thing I would change is the naming though. It seems like what's going on is we're listening for a hardware event and the user is still ultimately responsible for handling the event. I'd consider a rename. There are two separate actions going on here. One where we attach ourselves to listen for an event (async_receive) and another where we handle the event (async_read). I'd consider changing to something like **attach_receive_event**, **listen_for_receive**, etc and simply **uart_read**

Reply:
and then last question, can we still do it the freertos way if we'd like?


```
void rxCallback(){
   taskNotify(TASK_ONE_HANDLE, FLAG1);
}


void task_one(){
    while(1){
    // if no notification is received, freertos suspends this thread and receiveNotification never exits. This flag is not being polled by the user. 
// this only loops when the task is notified
     if (!receiveNotification(FLAG1)){
         continue;
    }

    // do smt...
}


}


```

Reply:
Thanks for sharing that Info, I wasn't familiar with the terminology, but it was helpful to learn that. I think using it this way with the RTOS should be possible. One key information is that calling `listen_for_receive` once will only give a call back once. So after receiving that notification, make sure to call the `listen_for_receive` (if you want to keep on listening). I think I will try to set one example with RTOS, so it will be clear to users.

Reply:
okay sounds good. there are lots of examples of deferred interrupt handling in freertos on their website. Using freertos would just not use async_receive and instead use `taskNotify()` (this is a freertos function). As long as we can do that it should be okay. Don't worry about creating working examples for every single use case